tests/FormTest.php

```diff
@@ -17,4 +21,56 @@ public function testGetField()
         $this->assertTrue($f->getField('test') instanceof \atk4\ui\FormField\Generic);
         $this->assertInstanceOf(\atk4\ui\FormField\Generic::class, $f->layout->getField('test'));
     }
+
+    public function testFormSubmit()
+    {
+        $f = new Form();
+        $f->app = new AppMockFT([
+            'catch_exceptions'        => false,
+            'always_run'              => false,
+            'catch_runaway_callbacks' => false,
+        ]);
+
+        $f->init();
+
+        $m = new Model();
+        $m->addField('name', ['default'=>'John']);
+        $m->addField('email', ['required'=>true]);
+        $m->addField('is_admin', ['default'=>false]);
+
+        $f->setModel($m, ['name', 'email']);
+
+        $this->assertEquals('John', $f->model->get('name'));
+
+        // fake some POST data
+        $_POST = ['atk_submit'=>'ajax', 'email'=>'john@yahoo.com', 'is_admin'=>'1'];
+
+        $submit_called = false;
+
+        $f->onSubmit(function ($f) use (&$submit_called) {
+
+            // field has default, but form didn't send value back
+            $this->assertEquals(null, $f->model['name']);
+
+            $this->assertEquals('john@yahoo.com', $f->model['email']);
+
+            // security check, unspecified field must not be changed
+            $this->assertEquals(false, $f->model['is_admin']);
+
+            $submit_called = true;
+        });
+
+        $f->render();
+        unset($f);
+
```

Comment:
I'd empty $_POST at the end of the test. In case there will be more tests in this class below, they could produce false results from the values still set to $_POST AFAIK.

Reply:
sure.

Although IIRC, phpunit does that too.